### PR TITLE
feat(migration): Diff local schema with dataset schema

### PR DESCRIPTION
**Pull request by lynnagara**
Parse DESCRIBE TABLE output into a schema which can be diffed with the
table writer schema.

Also fixes incorrectly logged errors in local migrations. These were being
incorrectly compared when there were default expressions or codecs in
the column definition.

### CODE REVIEW

File: snuba/datasets/schemas/__init__.py
Comment (fpacifici) [inline snuba/datasets/schemas/__init__.py]: Our migration system is so broken that it is hard to give a constructive feedback on how this should look like without 3 months of refactoring, still I think we can make a step in the right direction by start using ColumnType classes and subclasses instead of comparing strings.

The ColumnType abstraction is far from being a good one to manage migrations, but it is already present in the schema of the column and it would give you some structure at least.
So I suggest that, instead of fetching the ColumnType for the column from the schema, formatting it into string and comapring with a tweaked string coming from the `DESCRIBE` call, you parse the string coming from DESCRIBE into a ColumnType (that can be nested) and then compare it with the one present in the schema. 

Since this is not time critical now, I think it would be a good investment.

**Review by fpacifici — CHANGES_REQUESTED** [review]

File: snuba/datasets/schemas/__init__.py
Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: Tackled this first since I think having a reliable way to compare current database state with the schema of a dataset will be generally useful.

I think using our standard ColumnTypes (and doing away with MigrationSchemaColumn completely) will be good. Probably don't need a separate column type representation just for migration.

It seems like it will be quite a bit harder than comparing the strings though, since:
- We need to store another set of mappings to go from string column descriptions back into ColumnType structures. Going the other way is easier because ColumnType already implements the `for_schema()` method for this.
- There might be more than one way to represent columns in our schema, so it's ambiguous what the schema should actually be
- One case I can think of that is like this is our Nested column type where we use the `.` notation as a convention but it the same output can possibly be represented in multiple ways in the dataset

Comment (fpacifici) [reply snuba/datasets/schemas/__init__.py]: It seems both your first and second point apply equally to string comparison int hat you have to tweak the representations exactly in this method. I think we should think this problem through before adding a hack to avoid the warnings. 
Is there anything that would be blocked on this PR?

Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: No blockers. What do you mean that they apply equally though? My point was the opposite; I don't think those points apply to string comparison at all.
- Mapping from column type to string already exists (this is `for_schema()`), mapping from string to ColumnType needs to be added and is only needed if we are comparing the actual ColumnType instances
- There is a 1:many relationship between ColumnType: string output. Columns defined in schema always yield the same string in Clickhouse. The opposite isn't true, given a `DESCRIBE TABLE` output there are many ways it can be represented in the schema. (Nested vs Array is the one example of this)

Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: One other thing that is not very clear is how column type order is supposed to affect the query. For example is `WithCodec(WithDefault(x))` supposed to yield the same result as `WithDefault(WithCodec(x))`? Is one of these supposed to be invalid? Or are they meant to be smart enough to figure out what the right statement should be regardless of order applied?

Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: Last commit in this branch is the version with ColumnType and removed the MigrationSchemaColumn, there's a bit more code involved here

File: snuba/datasets/schemas/tables.py
Comment (lynnagara) [reply snuba/datasets/schemas/tables.py]: It would have caused circular dependencies since we needed to use it in snuba/datasets/schemas/__init__.py (in the previous version that did not remove it completely)

File: snuba/datasets/transactions.py
Comment (fpacifici) [inline snuba/datasets/transactions.py]: Why not just apply the alter without doing any check ?
It should be a noop if the column is already in the right format. But I would verify on Clickhouse if it complains bout this.

File: snuba/migrate.py
Comment (fpacifici) [inline snuba/migrate.py]: default_expr is not declared if the regex did not match anything. Did mypy complain ?

File: snuba/migrate.py
Comment (fpacifici) [inline snuba/migrate.py]: we need unit tests for this logic

File: snuba/migrate.py
Comment (fpacifici) [inline snuba/migrate.py]: Have you consider writing a grammar for this? It seems it would be easier to extend and manage than the ad hoc logic.
Like we did here  https://github.com/getsentry/snuba/blob/master/snuba/query/parser/expressions.py#L18-L43
It would save you all the manual regex matching, and all the custom logic to decide decide when to look for a modifier, which is very hard to verify.

File: snuba/clickhouse/columns.py
Comment (fpacifici) [inline snuba/clickhouse/columns.py]: This seems wrong now, We have a ColumnTypeWithModifier, but get_all_modifiers is outside of it, seems this is a leaky abstraction. I assume you did this to avoid depending on this file form errors_replacer, I guess for a circular dependency. Could you please expand on what the reason was ?

**Review by fpacifici — CHANGES_REQUESTED** [review]
I'd really invest some time in writing a proper grammar to parse the clickhouse type expressions. It would generate a much more maintainable and robust logic.

File: snuba/datasets/schemas/__init__.py
Comment (fpacifici) [reply snuba/datasets/schemas/__init__.py]:
```
For example is WithCodec(WithDefault(x)) supposed to yield the same statement as WithDefault(WithCodec(x))
```

That's a good question. I am not sure, could you please verify if clickhouse expects one form only ?

File: snuba/datasets/transactions.py
Comment (lynnagara) [reply snuba/datasets/transactions.py]: Do you mean making this potentially `LowCardinality(LowCardinality(String()))`? That won't work.

File: snuba/migrate.py
Comment (lynnagara) [reply snuba/migrate.py]: It just returns the untransformed default_expr if the regex doesn't match

File: snuba/migrate.py
Comment (lynnagara) [reply snuba/migrate.py]: I will give it a shot to see how it compares

File: snuba/clickhouse/columns.py
Comment (lynnagara) [reply snuba/clickhouse/columns.py]: The reason is doing `is_instance(x, ModifiedColumnType and LowCardinality in x.get_all_modifiers()` everywhere to figure out if a given column is low cardinality seems tedious. Given a column type it seems like it should be easier to find the list of modifiers. Would changing the implementation help so ColumnType doesn't reference ColumnTypeWithModifier help? I just did it as a straight copy/paste to get this working quickly for now.

File: snuba/datasets/schemas/__init__.py
Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: I tried both and only the first variant - `WithCodecs(WithDefault(x, def), codecs)`ends up getting formatted correctly by Clickhouse. But I guess my point wasn't about what happens today (which I'd expect to be not 100% correct) but what we should be doing if a user provides the reverse `WithDefault(withCodecs(x))`. Should we throw an error immediately? Or just accept it and transform it into the right thing? What if there are other examples out there that are both accepted by Clickhouse? Should we always enforce one "correct" order to apply the column types?

File: snuba/datasets/transactions.py
Comment (fpacifici) [reply snuba/datasets/transactions.py]: No, nevermind, I thought about running something like 
```
ALTER TABLE table MODIFY COLUMN {col_name} {col_type}
```
where col_type is the formatted expression from the Snuba schema (the source of truth), not from the current schema. But I realized you do not have access to the snuba schema here.

File: snuba/clickhouse/columns.py
Comment (fpacifici) [reply snuba/clickhouse/columns.py]: Can we then split this implementation? Let the general get_all_modifiers return [] and have a specific implementation for ColumnTypeWithModifier that provides the recursive implementation? 
That way ColumnTypeWithModufier actually has some responsibility and remove the isinstance calls.

File: snuba/datasets/schemas/__init__.py
Comment (fpacifici) [reply snuba/datasets/schemas/__init__.py]:
```
but what we should be doing if a user provides the reverse WithDefault(withCodecs(x)). Should we throw an error immediately? 
```
That's my question: does clickhouse accept both ? And if it accepts both does it consider them equivalent in a DESCRIBE TABLE statement ?

File: snuba/clickhouse/columns.py
Comment (lynnagara) [reply snuba/clickhouse/columns.py]: Yep, moved this change into https://github.com/getsentry/snuba/pull/803

File: snuba/datasets/schemas/__init__.py
Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: Clickhouse only accepts one. I suppose they are equivalent in that they are completely separate columns in the describe table output

**Review by fpacifici — COMMENTED** [review]
The grammar is a good step ahead, please add unit tests for that logic, it is easy to test since it is stateless and, since there are a lot of possible cases you can easily create many cases in this form:
https://github.com/getsentry/snuba/blob/master/tests/query/parser/test_expressions.py

Otherwise validating the grammar and avoiding regressions will be a pain due to the amount of cases to test.

File: snuba/migrate.py
Comment (lynnagara) [reply snuba/migrate.py]: Done, looks a lot clearer this way

File: snuba/migrate.py
Comment (lynnagara) [reply snuba/migrate.py]: Fixed. It doesn't cover some edge cases still (like absence of any ", " between pairs) but we shouldn't get invalid data in here and I haven't seen a concise way to accomplish this

**Comment by lynnagara** [comment]
@fpacifici Made your requested changes, any thoughts about this now?

**Comment by fpacifici** [comment]
I think this is still missing a good test coverage for the parser. I thought you were working on that. 
The issue I have in just  testing the difference between the local schema and the dataset schema is that it only tests the outcome of the migrations we happen to have now.
As a result we don't get a solid confidence that the grammar is right and it is properly processed in corner cases, let alone testing regressions on that grammar. Thus having a specific unit test for the grammar that test an extensive amount of cases (including corner cases) as we did for the query parsing. This gives you a good sense that the grammar is not regressing without depending on the specific dataset migrations we happen to have.

**Review by fpacifici — CHANGES_REQUESTED** [review]
back to you for test coverage

File: snuba/datasets/schemas/__init__.py
Comment (fpacifici) [inline snuba/datasets/schemas/__init__.py]: Any reason to compare the string representation in stead of the Type object now that we deal with ColumnType classes around ? I think we should turn this into a ColumnType comparison if it is possible.

File: snuba/migrations/migrate.py
Comment (fpacifici) [inline snuba/migrations/migrate.py]: Why this removal ? If the datasets has multiple schemas out of which several read ones and one writer (like matviews that have 2 read schemas and a write schema) the read schemas would not be migrated.

File: snuba/migrations/parse_schema.py
Comment (fpacifici) [inline snuba/migrations/parse_schema.py]: You should always leave an optional sequence of spaces after ( and before ) and  before and after commas. Clickhosue is probably not going to format its expressions by adding spaces but they would be valid expressions

File: snuba/migrations/parse_schema.py
Comment (fpacifici) [inline snuba/migrations/parse_schema.py]: There can be potentially multiple spaces after the comma

File: tests/migrations/test_parse_schema.py
Comment (fpacifici) [inline tests/migrations/test_parse_schema.py]: nice. thanks.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Almost there. I think there is an issue in the function that selects the schemas to migrate that now discards read schemas.

File: snuba/migrations/migrate.py
Comment (lynnagara) [reply snuba/migrations/migrate.py]: Yeah good point, I think this was an oversight.

**Review by fpacifici — APPROVED** [review]

**Comment by fpacifici** [comment]
Please fix the test and see the other comments from my previous reviews before merging

File: snuba/migrations/parse_schema.py
Comment (lynnagara) [reply snuba/migrations/parse_schema.py]: Makes sense, done

File: snuba/datasets/schemas/__init__.py
Comment (lynnagara) [reply snuba/datasets/schemas/__init__.py]: Not really, done

**Pull request merged**
